### DISCUSSION TITLE
Tycho: requires 'osgi.ee; (&(osgi.ee=JavaSE)(version=11))' but it could not be found

### DISCUSSION THREAD

**Discussion by dwagelaar** [eclipse-tycho/tycho#554] [Q&A] [answered]
In order to build ATL against the latest Eclipse release plug-ins, I've started the effort to port it to JavaSE-11 (https://bugs.eclipse.org/bugs/show_bug.cgi?id=576366). Only a few changes were required in the source code, and I got ATL to build fairly quickly within the Eclipse IDE.

The tycho build is another matter, however. When updating to the latest tycho release (required for JavaSE-11), it spits out the following error:

```
[INFO] Computing target platform for MavenProject: org.eclipse.m2m.atl:org.eclipse.m2m.atl.emftvm:4.6.0-SNAPSHOT @ /home/dennis/Documents/Werk/Projects/org.eclipse.atl/plugins/org.eclipse.m2m.atl.emftvm/pom.xml
[ERROR] Cannot resolve target definition:
[ERROR]   Software being installed: org.eclipse.emf.sdk.feature.group 2.28.0.v20211110-0654
[ERROR]   Missing requirement: org.eclipse.core.jobs 3.12.0.v20210723-1034 requires 'osgi.ee; (&(osgi.ee=JavaSE)(version=11))' but it could not be found
[ERROR]   Cannot satisfy dependency: org.eclipse.core.runtime 3.24.0.v20210910-0750 depends on: osgi.bundle; org.eclipse.core.jobs [3.12.0,4.0.0)
[ERROR]   Cannot satisfy dependency: org.eclipse.emf.common.feature.group 2.24.0.v20210924-1718 depends on: org.eclipse.equinox.p2.iu; org.eclipse.core.runtime 0.0.0
[ERROR]   Cannot satisfy dependency: org.eclipse.emf.feature.group 2.28.0.v20211110-0654 depends on: org.eclipse.equinox.p2.iu; org.eclipse.emf.common.feature.group [2.24.0.v20210924-1718,2.24.0.v20210924-1718]
[ERROR]   Cannot satisfy dependency: org.eclipse.emf.sdk.feature.group 2.28.0.v20211110-0654 depends on: org.eclipse.equinox.p2.iu; org.eclipse.emf.feature.group [2.28.0.v20211110-0654,2.28.0.v20211110-0654]
[ERROR] 
[ERROR] Failed to resolve target definition /home/dennis/Documents/Werk/Projects/org.eclipse.atl/releng/org.eclipse.m2m.atl.releng.parent/tp/org.eclipse.m2m.atl.tp.target: See log for details -> [Help 1]
```

https://wiki.eclipse.org/Tycho/Dependency_Resolution_Troubleshooting suggests that the target definition is inconsistent. I've since been experimenting with target platform definitions, and adding "JavaSE-11" as targetJRE seems to solve the dependency resolution problems, but then switches the Java compiler to version 1.5 source level (`@Override` on interface methods, multi-catch not supported):

```
[ERROR] Failed to execute goal org.eclipse.tycho:tycho-compiler-plugin:2.5.0:compile (default-compile) on project org.eclipse.m2m.atl.emftvm: Compilation failure: Compilation failure: 
[ERROR] /home/dennis/Documents/Werk/Projects/org.eclipse.atl/plugins/org.eclipse.m2m.atl.emftvm/src/org/eclipse/m2m/atl/emftvm/util/LazyOrderedSet.java:[1891] 
[ERROR] 	public void add(final int index, final E element) {
[ERROR] 	            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[ERROR] The method add(int, E) of type LazyOrderedSet<E> must override a superclass method
...
[ERROR] /home/dennis/Documents/Werk/Projects/org.eclipse.atl/plugins/org.eclipse.m2m.atl.emftvm/src/org/eclipse/m2m/atl/emftvm/impl/EmftvmFactoryImpl.java:[1662] 
[ERROR] 	} catch (InstantiationException | IllegalAccessException | IllegalArgumentException | InvocationTargetException
[ERROR] 				| NoSuchMethodException | SecurityException | ClassNotFoundException e) {
[ERROR] 	         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[ERROR] Multi-catch parameters are not allowed for source level below 1.7
```

The PR can be found at https://git.eclipse.org/r/c/mmt/org.eclipse.atl/+/189434

Does anyone have any idea what is going on?

P.S. I found https://github.com/eclipse/tycho/issues/205, but that removing the target definition is the starting situation in this case.

**Comment by laeubi** [comment]
I noticed you are using a combination of a target+repository references, this makes it a bit hard to follow whats going on, is there are strong reason to do so?

The source level has nothing to do with the target platform but might be derived from your bundles BREE, can you check that all your bundles are using JavaSe-11 as BREE, have the correct JRE set in the classpath and your build is using a JDK11 or is configured with a corresponding one in toolchains.xml?

**Reply by dwagelaar** [reply]
Thank you for your reply!

- There's no reason to use targets and repositories in combination: AFAIK I moved repositories into a (disabled) Maven profile, and use only targets in the Gerrit PR.
- I've gone over the BREE's already, as I suspected issues there. I will do this again, and perhaps also look at the Orbit dependencies, and whether excluding them from the features makes a difference...

**Comment by mickaelistria** [comment]
>  adding "JavaSE-11" as targetJRE seems to solve the dependency resolution problems, but then switches the Java compiler to version 1.5 source level

This is the issue you should look at carefully. There is no reason for this to happen and I suspect it's an issue in your end. Check you build files for anything that would enforce a 1.5 source level, it's not default Tycho behavior.

**Reply by dwagelaar** [reply]
Thank you for your reply! I will look outside tycho then (e.g. BREE), and for the moment assume that the tycho configuration is working...

**Reply by laeubi** [reply]
Running with -X could also reveal some interesting things.

**Reply by laeubi** [reply]
I think this could be related to https://github.com/eclipse/tycho/issues/598 are you running on a mac?

**Comment by dwagelaar** [comment]
No, I'm running on Fedora Linux 34, but the build also fails on
https://ci.eclipse.org/atl/job/mmt-atl-pr/84/console

**Comment by dwagelaar** [comment, marked as answer]
ATL has since migrated to github, and I've had to revisit the Jenkins build pipeline. In this context, I've managed to get ATL to build using Tycho 4.0.8, such that it can be built against the latest Eclipse codebase.

See also:
- https://github.com/eclipse-atl/atl/issues/4
- https://github.com/eclipse-atl/atl/commit/4dee212db289d74c5337ec5f1e0de4f9108ecdbc

I've added a target platform definition with an explicit targetJRE to get around the problem of having various bundles in the dependency graph that require a lower version JRE than its upstream dependencies. This is typical for OSGi bundles that want to be able to build against older versions of the Eclipse codebase.
